libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -133,52 +163,217 @@ void ContractLevelChecker::findDuplicateDefinitions(map<string, vector<T>> const
 	}
 }
 
+vector<ContractDefinition const*> resolveBaseContracts(ContractDefinition const& _contract)
+{
+	vector<ContractDefinition const*> resolvedContracts{_contract.baseContracts().size(), nullptr};
+
+	transform(
+		_contract.baseContracts().cbegin(),
+		_contract.baseContracts().cend(),
+		resolvedContracts.begin(),
+		[] (ASTPointer<InheritanceSpecifier> const& _specifier)
+		{
+			Declaration const* baseDecl =
+				_specifier->name().annotation().referencedDeclaration;
+			return dynamic_cast<ContractDefinition const*>(baseDecl);
+		}
+	);
+
+	return resolvedContracts;
+}
+
+ContractLevelChecker::FunctionInfo ContractLevelChecker::findAndCheckOverride(
+	ContractDefinition const& _derived,
+	FunctionDefinition const& _function,
+	ContractDefinition const& _base
+)
+{
+	// Look for possible override in all functions of _base
+	for (FunctionDefinition const* baseFunc: _base.definedFunctions())
+	{
+		if (!hasEqualNameAndParameters(_function, *baseFunc))
+			continue;
+
+		if (!checkFunctionOverride(_function, *baseFunc))
```

Comment:
I think I would find it easier to understand if the checks were not part of the recursion.